architecture/2024-08-applicationdeployment-controller.md

```diff
@@ -0,0 +1,370 @@
+align these sections with the product's layers. This will help readers use
+their current understanding to understand your ideas.
+
+Discuss the rationale behind architectural choices and alternative options 
+considered during the design process.
+-->
+
+#### ApplicationDeployment Custom Resource Definition
+
+The `ApplicationDeployment` resource will be a new Kubernetes CRD that will be used to contain the Bicep template and parameters for deploying Radius resources. The CRD will have the following fields:
+
+- `spec.template`: The Bicep template file that defines the resources to deploy
+- `spec.parameters`: The Bicep parameters file that defines the parameters for the template
+- `status`: The status of the deployment, which will be one of the following values:
+  - `Pending`: The deployment is pending
+  - `Deployed`: The deployment has been completed
+  - `Missing Dependencies`: The deployment is missing dependencies
+
+#### `rad applicationdeployment generate` Command
+
+The `rad applicationdeployment generate` command will be a new command that will generate an ApplicationDeployment resource from a Bicep template and parameters file. The command will take the following arguments:
+
+- `-t, --template`: The path to the Bicep template file
+- `-p, --parameters`: The path to the Bicep parameters file
+- `-o, --output`: The path to the output file where the ApplicationDeployment resource will be written
+
+The command will read the Bicep template and parameters files and generate an ApplicationDeployment resource with the template and parameters set. The command will write the ApplicationDeployment resource to the output file.
+
+#### ApplicationDeployment Controller
+
+The `radius-controller` will be updated to include a new controller that watches for ApplicationDeployment resources. When an ApplicationDeployment resource is created or updated, the controller will read the Bicep template and parameters from the ApplicationDeployment resource and send them to the `radius-bicep` service. The `radius-bicep` service will compile the Bicep and bicepparam files into ARM JSON and return the ARM JSON to the `radius-controller`.
+
+For each of the resources in the template, the `radius-controller` will determine if the dependencies for the resource already exist by querying UCP. If they do not, then the controller will mark the ApplicationDeployment resource as "Missing Dependencies" with the resource that is missing and will not deploy the resource. If the dependencies do already exist for each of the resources in the template, then the controller will send the ARM JSON payload to UCP to deploy the resources. Once the deployment is complete, The controller will trigger a re-deployment of resources with status "Missing Dependencies" that depend on the newly deployed resources. The controller will update the status of the ApplicationDeployment resource to "Deployed" once all resources have been deployed.
```

Comment:
I think the design can be simpler than this. 

> For each of the resources in the template, the `radius-controller` will **determine if the dependencies for the resource already exist by querying UCP**.

If we needed to do this it would actually be really hard. It's hard to just look at a Bicep template and compute exactly the resource IDs of the resources.